<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.6: http://docutils.sourceforge.net/" />
<title>Detailed papers on LAST</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 5951 2009-05-18 18:03:10Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left{
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: left }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
<style type="text/css">

/* Style sheet for LAST HTML documents */
h1 { color: navy }
h2 { color: teal }
div.document { margin-left: auto; margin-right: auto; max-width: 45em }
strong { color: red }
.option-list td { padding-bottom: 1em }
table.field-list { border: thin solid green }

</style>
</head>
<body>
<div class="document" id="detailed-papers-on-last">
<h1 class="title">Detailed papers on LAST</h1>

<p>LAST has many ingredients, some of which are described in these
papers.  If you find an ingredient useful, please cite the
corresponding paper.  Citation is important because it provides
feedback on which research work was useful, and helps to justify the
research to society.</p>
<ol class="arabic">
<li><p class="first"><a class="reference external" href="http://genome.cshlp.org/content/21/3/487.long">Adaptive seeds tame genomic sequence comparison</a>.  Kie≈Çbasa SM,
Wan R, Sato K, Horton P, Frith MC.  Genome Res. 2011 21(3):487-93.</p>
<p>This describes the main algorithms used by LAST.</p>
</li>
<li><p class="first"><a class="reference external" href="http://nar.oxfordjournals.org/content/38/7/e100.long">Incorporating sequence quality data into alignment improves DNA
read mapping</a>.  Frith MC, Wan R, Horton P.  Nucleic Acids
Res. 2010 38(7):e100.</p>
<p>How LAST uses sequence quality data.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.biomedcentral.com/1471-2105/11/80">Parameters for Accurate Genome Alignment</a>.  Frith MC, Hamada M,
Horton P.  BMC Bioinformatics. 2010 11:80.</p>
<p>Choice of score parameters, ambiguity of alignment columns, and
gamma-centroid alignment.</p>
</li>
<li><p class="first"><a class="reference external" href="http://nar.oxfordjournals.org/content/39/4/e23.long">A new repeat-masking method enables specific detection of
homologous sequences</a>.  Frith MC.  Nucleic Acids Res. 2011
39(4):e23.</p>
<p>This describes the tantan algorithm for finding simple /
low-complexity / tandem repeats, which reliably prevents
non-homologous alignments, unlike other repeat finders.</p>
</li>
<li><p class="first"><a class="reference external" href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0028819">Gentle masking of low-complexity sequences improves homology
search</a>.  Frith MC.  PLoS One. 2011 6(12):e28819.</p>
<p>This describes what LAST does with repeats after they have been
found.</p>
</li>
<li><p class="first"><a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/27/22/3085.long">Probabilistic alignments with quality scores: an application to
short-read mapping toward accurate SNP/indel detection</a>.  Hamada
M, Wijaya E, Frith MC, Asai K.  Bioinformatics. 2011
27(22):3085-92.</p>
<p>Describes probabilistic alignment using sequence quality data, and
LAMA alignment.</p>
</li>
<li><p class="first"><a class="reference external" href="http://nar.oxfordjournals.org/content/40/13/e100.long">A mostly traditional approach improves alignment of
bisulfite-converted DNA</a>.  Frith MC, Mori R, Asai K.  Nucleic
Acids Res. 2012 40(13):e100.</p>
<p>This describes alignment of bisulfite-converted DNA, and an update
for use of fastq quality data that allows for non-uniform base
frequencies.</p>
</li>
<li><p class="first"><a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/29/8/965.long">An approximate Bayesian approach for mapping paired-end DNA reads
to a reference genome</a>.  Shrestha AM, Frith MC.
Bioinformatics. 2013 29(8):965-72.</p>
<p>This describes the algorithm used by last-pair-probs.</p>
</li>
<li><p class="first"><a class="reference external" href="http://nar.oxfordjournals.org/content/42/7/e59.long">Improved search heuristics find 20,000 new alignments between
human and mouse genomes</a>.  Frith MC, No√© L.  Nucleic Acids
Res. 2014 42(7):e59.</p>
<p>This describes sensitive DNA seeding (MAM8 and MAM4).</p>
</li>
<li><p class="first"><a class="reference external" href="http://bioinformatics.oxfordjournals.org/content/30/24/3575.long">Frameshift alignment: statistics and post-genomic
applications</a>.  Sheetlin SL, Park Y, Frith MC, Spouge JL.
Bioinformatics. 2014 30(24):3575-82.</p>
<p>Describes DNA-versus-protein alignment allowing for frameshifts.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.genomebiology.com/content/16/1/106">Split-alignment of genomes finds orthologies more accurately</a>.
Frith MC, Kawaguchi R.  Genome Biology. 2015 16:106.</p>
<p>Describes the split alignment algorithm, and its application to
whole genome alignment.</p>
</li>
<li><p class="first"><a class="reference external" href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btw742">Training alignment parameters for arbitrary sequencers with
LAST-TRAIN</a>.  Hamada M, Ono Y, Asai K Frith MC.
Bioinformatics. 2017 33(6):926-928.</p>
<p>Describes last-train.</p>
</li>
<li><p class="first"><a class="reference external" href="https://ieeexplore.ieee.org/document/8288582/">A Simplified Description of Child Tables for Sequence Similarity
Search</a>.  Frith MC, Shrestha A.  IEEE/ACM Trans Comput Biol
Bioinform. 2018.</p>
<p>Describes how LAST uses child tables.</p>
</li>
</ol>
<div class="section" id="external-methods">
<h2>External methods</h2>
<p>LAST of course owes its ideas to much previous research.  Here are
listed only implementations that are directly used in LAST.</p>
<ul>
<li><p class="first"><a class="reference external" href="http://nar.oxfordjournals.org/content/33/15/4987.long">The Gumbel pre-factor k for gapped local alignment can be estimated
from simulations of global alignment</a>.  Sheetlin S, Park Y, Spouge
JL.  Nucleic Acids Res. 2005 33(15):4987-94.</p>
<p>Describes how E-values are calculated.</p>
</li>
<li><p class="first"><a class="reference external" href="http://www.biomedcentral.com/1756-0500/5/286">New finite-size correction for local alignment score
distributions</a>.  Park Y, Sheetlin S, Ma N, Madden TL, Spouge JL.
BMC Res Notes. 2012 5:286.</p>
<p>Describes a correction that makes the E-values more accurate for
short sequences.</p>
</li>
<li><p class="first"><a class="reference external" href="https://www.usenix.org/publications/login/february-2011-volume-36-number-1/gnu-parallel-command-line-power-tool">GNU Parallel - The Command-Line Power Tool</a>.  Tange O.  ;login:
The USENIX Magazine. 2011:42-47.</p>
<p>It seems traditional not to cite this kind of ingredient, which is
unfortunate because the same reasons for citation apply.</p>
</li>
</ul>
</div>
</div>
</body>
</html>
